### ISSUE TITLE
API

### ISSUE THREAD

**Issue by goshawk22** [dabreegster/abstreet#245]
Would it be possible to implement and api to control the simulation? This way further projects could be based off this to improve road infrastructure.

**Comment by dabreegster** [comment]
Definitely. A few people have brought the idea up. I could see something for running the simulation without any UI and extracting analytics in order to do reinforcement learning. I've also heard the idea of running the game in a presentation/exhibit mode and being able to script camera movements and such. Do you have anything specific in mind? I could fulfill this feature request best with a specific use case in mind.

**Comment by goshawk22** [comment]
I thought it would be useful for reinforcement learning, but I guess a general api would be the best option.

**Comment by dabreegster** [comment]
I haven't done RL since a class ages ago, so I'm a little fuzzy what this would look like. Maybe traffic signal timing is a good first example to think about. The action would be editing signals, then the observation would be to run the simulation for some amount of time (maybe the full 24 hours if things are going well, but short-circuit if gridlock is detected). Data already visible in the game can be exported through an API, like [intersection delay](https://github.com/dabreegster/abstreet/blob/3af929f83069dd27df29d51e8f22e4176ddf0a29/sim/src/analytics.rs#L33). Specifying the signal edits can already be done through a [JSON file](https://github.com/dabreegster/seattle_traffic_signals/blob/master/data/1635790583.json), but actually coming up with this thing is pretty hard from scratch. It would make sense to expose some of the methods [here](https://github.com/dabreegster/abstreet/blob/master/map_model/src/objects/traffic_signals.rs) and elsewhere. (Signal editing is complicated; maybe editing lane types is a simpler start.)

Could you write some simple pseudocode to describe one of the particular RL things you'd like to try? I'm really use-case driven for designing APIs; the potential API surface to control everything in the game is absolutely massive, and I think it's hard to evaluate whether a potential API is useful or not without having examples in mind.

**Comment by RestitutorOrbis** [comment]
I'm not `goshawk22` but I think an API has value beyond RL. For example, making A/B Street a web application may be a bit difficult if A/B Street is entirely run client-side as that may place too much load on the user's machine. In this case, it might be worth it to extract an API that the server could rely upon to run A/B Street headlessly to determine what to send to the client side which would then use another API from A/B Street to actually draw an image the user could view.

**Comment by dabreegster** [comment]
Definitely a possibility, but an API feels like a solution in search of a problem. When the web support initially lands, there are a bunch of techniques to speed up simulation for slower clients, and some kind of API is definitely one of them. In the meantime, any specific things we could do with an API now?

**Comment by RestitutorOrbis** [comment]
> Definitely a possibility, but an API feels like a solution in search of a problem. When the web support initially lands, there are a bunch of techniques to speed up simulation for slower clients, and some kind of API is definitely one of them. In the meantime, any specific things we could do with an API now?

That's a fair assessment. In that case, the most relevant usecase for an API for me as a developer would be integration testing. Currently, A/B Street doesn't have any tests (as far as I know, correct me if I'm wrong) that validate certain properties of the system hold after changes. Things like validating that gridlock is avoided in a certain scenario remains impossible/unlikely after a change to the relevant code would (I imagine) be useful.

**Comment by dabreegster** [comment]
The `game` crate has a `--prebake` that runs `montlake` and `lakeslice` to completion, the two maps that (after fighting very very hard) complete fully. I recently added `--smoketest` too, to run a few hours on every map and make sure things like spawning at borders don't outright crash. More extensive integration testing should definitely happen. Before I started this project, I remember reflecting on my college simulation project and thinking about very small unit tests for pieces of sim behavior, like the exact mechanics of lane-changing and parking. But that... never happened. :)

An API focused around testing is definitely one thing to think about short-term. Would there be use in having the test runner be in something other than Rust, necessitating an RPC-based API?

**Comment by dabreegster** [comment]
If you're curious, there were some more interesting integrationish tests at some point: https://github.com/dabreegster/abstreet/blob/15f6981b1de4fb4346b044a7485cf41b01cc8fd5/tests/src/trips.rs
But I didn't maintain them, because they didn't seem to add much value/confidence. Definitely worth trying again.

**Comment by RestitutorOrbis** [comment]
I think the value of an RPC based API is relevant towards increasing the number of possible contributors. I personally find Rust to be significantly more difficult than programming in Python and this is completely reasonable for most of what A/B Street does but I think there is value in having tests written in an easier language like Python and this would open opportunities for contributors who are unfamiliar with Rust. That being said, I don't think an RPC approach is absolutely necessary to achieve this as things like https://github.com/PyO3/pyo3 exist.

**Comment by dabreegster** [comment]
https://dabreegster.github.io/abstreet/api.html

There's now a more fleshed-out JSON API, with an example Python script that runs a simple experiment to fiddle with signal timing. I'm going to add a few more things to the API to extract data like throughput. But beyond that, it'll be use case driven.
